stb_image_write.h

```diff
@@ -1062,23 +1069,32 @@ static void stbiw__encode_png_line(unsigned char *pixels, int stride_bytes, int
       }
    }
    switch (type) {
-      case 1: for (i=n; i < width*n; ++i) line_buffer[i] = z[i] - z[i-n]; break;
-      case 2: for (i=n; i < width*n; ++i) line_buffer[i] = z[i] - z[i-signed_stride]; break;
-      case 3: for (i=n; i < width*n; ++i) line_buffer[i] = z[i] - ((z[i-n] + z[i-signed_stride])>>1); break;
-      case 4: for (i=n; i < width*n; ++i) line_buffer[i] = z[i] - stbiw__paeth(z[i-n], z[i-signed_stride], z[i-signed_stride-n]); break;
-      case 5: for (i=n; i < width*n; ++i) line_buffer[i] = z[i] - (z[i-n]>>1); break;
-      case 6: for (i=n; i < width*n; ++i) line_buffer[i] = z[i] - stbiw__paeth(z[i-n], 0,0); break;
+      case 1: for (i=pixel_bytes; i < width*n; ++i) line_buffer[i] = z[i] - z[i-pixel_bytes]; break;
+      case 2: for (i=pixel_bytes; i < width*n; ++i) line_buffer[i] = z[i] - z[i-signed_stride]; break;
+      case 3: for (i=pixel_bytes; i < width*n; ++i) line_buffer[i] = z[i] - ((z[i-pixel_bytes] + z[i-signed_stride])>>1); break;
+      case 4: for (i=pixel_bytes; i < width*n; ++i) line_buffer[i] = z[i] - stbiw__paeth(z[i-pixel_bytes], z[i-signed_stride], z[i-signed_stride-pixel_bytes]); break;
+      case 5: for (i=pixel_bytes; i < width*n; ++i) line_buffer[i] = z[i] - (z[i-pixel_bytes>>1]); break;
```

Comment:
@WearerOfShoes, I'm very interested in this, too...

Reply:
@nothings: So it took me a while to figure out, but apart from the right-shift bug you pointed out, the code is actually correct with regards to `width * n` vs  `width * pixel_bytes` in the for-loops because `width` has already been multiplied by `comp_bytes` (i.e. 1 for 8-bit PNGs, 2 for 16-bit PNGs) in stbiw__write_png_to_mem_core():

```
   // all calculations in PNG are based on bytes in 16-bit mode,
   // so just treated it as a twice-as-wide 8-bit image works in most cases
   // exceptions are patches up by sending along comp_bytes
   x *= comp_bytes;
```

I also ran the tests in image_write_test and all PNG generated files are seemingly correct.

(BTW, I can make a new PR based on a fork of @WearerOfShoes work so we can finish this if you like)